tests/integration/test_recipe.py

```diff
@@ -2199,7 +2199,13 @@ def test_landmask_no_fx(tmp_path, patched_failing_datafinder, config_user):
         fx_variables = product.settings['add_fx_variables']['fx_variables']
         assert isinstance(fx_variables, dict)
         fx_variables = fx_variables.values()
-        assert not any(fx_variables)
+        # usable fx data is found for CMIP6 in piControl
+        # the test is to check if all those found instances are only CMIP6's
+        if fx_variables:
+            count_cmip6 = [
+                p for p in fx_variables if p.get('project', None) == 'CMIP6'
+            ]
+            assert len(fx_variables) == len(count_cmip6)
```

Comment:
I'm not sure why this change is necessary. The `patched_failing_datafinder` that is used your is not dependent on the `exp` right now, so your changes here shouldn't affect the outcome of this change. Can you double-check?

Reply:
yes I double checked - the assertion fails for CMIP6 since it finds fx files in piControl, with it passing for CMIP5 since there were no fx files found there

Reply:
If this is the case then I think something weird happens in your new code or the test - the failing data finder is designed to fail in this situation, regardless of `exp`...

Reply:
the original test was `assert not any(fx_variables)` which was passing since there were no fx files for either CMIP5 or CMIP6; now, with the lookup in piControl, it finds the piControl fx file so the original test was failing for CMIP6, but still passing for CMIP5, nonetheless it was failing at exit

Reply:
But due the use of the [`patched_failing_datafinder`](https://github.com/ESMValGroup/ESMValCore/blob/d75eb9102bf1938bb33da2034c8a25973651532d/tests/integration/test_recipe.py#L279-L306), this test shouldn't find any data, regardless of CMIP5 or CMIP6 or the `exp` of them! I'm really confused why it suddenly finds data for `piControl` now...

Reply:
gah! blast, you're right, it finds whatever you want it to find, it did find `historical-222` instead of `piControl`  :man_facepalming:

Reply:
it finds the piControl data all too well, man! Here's the output if I force fail the test:
```
dict_values([])
dict_values([{'preprocessor': 'landmask', 'project': 'CMIP6', 'mip': 'Lmon', 'exp': 'piControl', 'start_year': 2000, 'end_year': 2005, 'ensemble': 'r1i1p1f1', 'variable_group': 'sftof', 'short_name': 'sftof', 'diagnostic': 'diagnostic_name', 'dataset': 'CanESM5', 'grid': 'gn', 'recipe_dataset_index': 1, 'institute': ['CCCma'], 'activity': 'CMIP', 'alias': 'CMIP6', 'original_short_name': 'sftof', 'standard_name': 'sea_area_fraction', 'long_name': 'Sea Area Percentage', 'units': '%', 'modeling_realm': ['ocean'], 'frequency': 'fx', 'filename': '/tmp/pytest-of-valeriu/pytest-96/popen-gw0/test_landmask_no_fx0/input/sftof_Lmon_CanESM5_piControl_r1i1p1f1_gn_1990_1999.nc'}])
--------------------------------------------------------------------- Captured log call ----------------------------------------------------------------------
WARNING  esmvalcore._recipe:_recipe.py:455 Missing data for fx variable 'sftlf'
WARNING  esmvalcore._recipe:_recipe.py:455 Missing data for fx variable 'sftof'
WARNING  esmvalcore._recipe:_recipe.py:441 Missing data for fx variable sftlf and experiment historical; since data is CMIP6, will try experiment piControl, most fx stuff is there.
WARNING  esmvalcore._recipe:_recipe.py:450 Failed to find data for CMIP6 fx variable sftlf with experiment piControl as well.
WARNING  esmvalcore._recipe:_recipe.py:441 Missing data for fx variable sftof and experiment historical; since data is CMIP6, will try experiment piControl, most fx stuff is there.
WARNING  esmvalcore._recipe:_recipe.py:455 Missing data for fx variable 'sftlf'
```
exactly what I want :grin:

Reply:
Not exactly what I want :smile: : `sftof_Lmon_CanESM5_piControl_r1i1p1f1_gn_1990_1999.nc`. There is no `sftof` in `Lmon`.

The main problem here is that `fx_info['mip']` is set to `Lmon` after the automatic mip search, which causes the data finder to not fail (since the `mip` is not `fx` anymore). In this case you need to repeat the following logic (i.e. the automatic mip search) also for the new `exp` (after removing `mip` again from `fx_info` if necessary):

https://github.com/ESMValGroup/ESMValCore/blob/d75eb9102bf1938bb33da2034c8a25973651532d/esmvalcore/_recipe.py#L418-L433

Reply:
aye! but why is Lmon added as a valid mip for sftof in the first place? All my new bit of code is doing is jiggling the exp for a set of otherwise fixed and hopefully valid data parameters?

Reply:
That's why:

https://github.com/ESMValGroup/ESMValCore/blob/d75eb9102bf1938bb33da2034c8a25973651532d/esmvalcore/_recipe.py#L389-L391

I think this was necessary since code at a later stage needs the `mip` attribute.

Reply:
OK but this is not good since we need to set the correct data/file parameters right before we call  `_get_input_files(fx_info, config_user)[0]` - why are we running that function with invalid data descriptors, like sftof-Lmon?

Reply:
We didn't do that before the changes here :wink: As I said, you need to make sure to use same logic as for the original `exp`: If `mip` exists -> use this; if not -> call `_search_fx_mip`

Reply:
as in - the moment we call the data finder function we need to make sure that `fx_info` dict is correct and not massage it afterwards if we don't find files, unless one changes the exp to piControl which is a CMOR table-independent parameter (like I'm doing here) - what am I missing here?

Reply:
arite, that whole block needs to be a function on its own, and an annoying one too, since that needs to be called once (ideally) and the `fx_info` needs to be set correctly before calling the data finder file fetter bleh :grin:

Reply:
OK Manu, I have made that mip jiggling a function now, that gets applied before we look for piControl, rehabilitated the old test that needed to stay the same and added two failsafe tests that should fail without using the `patched_datafinder` - these cover the new functionality with piControl, but they cover it synthetically, I don't want to start using yaml files containing actual paths to filesso I can write a test that doesn't find eg historical but finds piControl, that's overkill for this PR, what do you think?